### PR TITLE
Adds support for metadata values of type array

### PR DESCRIPTION
**Pull request by jgpruitt**
Metadata values containing jsonb arrays are now supported. Predicates can take a list as a value. A comparison operator of "@>" now tests for array containment. This is an indexed operation.

```python
await vec.upsert([
    (uuid.uuid4(), '''{"key0": [1,2,3,4]}''', "the brown fox", [1.0, 1.8]),
    (uuid.uuid4(), '''{"key0": [5,6,7], "key3": 3}''', "the brown fox", [1.0, 1.8]),
])
```

```python
rec = await vec.search([1.0, 2.0], limit=4, predicates=Predicates("key0", "@>", [1, 2]))
assert len(rec) == 1
rec = await vec.search([1.0, 2.0], limit=4, predicates=Predicates("key0", "@>", [3, 7]))
assert len(rec) == 0
rec = await vec.search([1.0, 2.0], limit=4, predicates=Predicates("key0", "@>", [9]))
assert len(rec) == 0
rec = await vec.search([1.0, 2.0], limit=4, predicates=Predicates("key0", "@>", [4]))
assert len(rec) == 1

rec = await vec.search([1.0, 2.0], limit=4, predicates=Predicates("key0", "@>", [6,7]) and Predicates("key3","==", 3))
assert len(rec) == 1
rec = await vec.search([1.0, 2.0], limit=4, predicates=Predicates("key0", "@>", [6,7]) and Predicates("key3","==", 6))
assert len(rec) == 0
```

### CODE REVIEW

File: timescale_vector/client.py [hunk 1]
```diff
@@ -433,13 +434,29 @@ def build_query(self, params: List) -> Tuple[str, List]:
                     params.append(value)
                     continue
 
+                # if array containment
+                if operator == "@>" and isinstance(value, list):
+                    field_cast = ''
+                    if len(value) == 0:
```
Comment (cevian) [inline timescale_vector/client.py:437]: what does this even mean? shouldn't an empty list be an error?

File: timescale_vector/client.py [hunk 2]
```diff
@@ -433,13 +434,29 @@ def build_query(self, params: List) -> Tuple[str, List]:
                     params.append(value)
                     continue
 
+                # if array containment
+                if operator == "@>" and isinstance(value, list):
+                    field_cast = ''
```
Comment (cevian) [inline timescale_vector/client.py]: how are string values handled? Should strings be tested for below? or should the default cast be to text[]?

Comment (cevian) [reply timescale_vector/client.py]: we should probably add a test for string-typed value array

**Review by cevian — CHANGES_REQUESTED** [review]
Minor changes requested

File: timescale_vector/client.py (hunk 1 above)
Comment (jgpruitt) [reply timescale_vector/client.py:437]: Oddly enough, an empty array seems to match everything. Yeah, that's probably not desirable.
image

File: timescale_vector/client.py (hunk 2 above)
Comment (jgpruitt) [reply timescale_vector/client.py]: good catch.
actually, i just realized that I can make the code much more flexible and simpler...

Comment (jgpruitt) [reply timescale_vector/client.py]: latest push allows for PredicateValues that are tuples or lists containing values of mixed types

File: timescale_vector/client.py (hunk 1 above)
Comment (jgpruitt) [reply timescale_vector/client.py:437]: latest commit throws ValueError on empty list/tuple

File: timescale_vector/client.py [hunk 3]
```diff
@@ -431,18 +432,24 @@ def build_query(self, params: List) -> Tuple[str, List]:
...
-                    field_cast = '::timestamptz'   
-
-                where_conditions.append(f"(metadata->>'{field}'){field_cast} {operator} {param_name}")
-                params.append(value) 
+                
+                elif operator == "@>" and (isinstance(value, list) or isinstance(value, tuple)):
+                    if len(value) == 0:
+                        raise ValueError("Invalid value. Empty lists and empty tuples are not supported.")
+                    json_value = json.dumps(value)
+                    where_conditions.append(f"metadata @> jsonb_build_object('{field}', {param_name}::jsonb)")
```
Comment (cevian) [inline timescale_vector/client.py:440]: Did you verify this is still an indexed operation?

**Review by cevian — APPROVED** [review]

File: timescale_vector/client.py (hunk 3 above)
Comment (jgpruitt) [reply timescale_vector/client.py:440]: yes. it is

Comment (jgpruitt) [reply timescale_vector/client.py:440]:
```sql
create table bob (id uuid, metadata jsonb);

insert into bob (id, metadata)
select
  gen_random_uuid()
, jsonb_build_object
  ( 'x', x
  , 'y', array_sample(array['A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K'], 3)
  )
from generate_series(1, 10000) x
;

create index on bob using gin (metadata jsonb_path_ops);
```
image

image

**Review by cevian — APPROVED** [review]
LGTM. will talk to you about how to run the nbdev commands

**Pull request merged**